rust/pact_matching_ffi/src/lib.rs

```diff
@@ -9,3 +9,143 @@ pub mod error;
 pub mod log;
 pub mod models;
 pub(crate) mod util;
+
+use crate::util::*;
+use libc::c_char;
+use models::message::Message;
+use pact_matching as pm;
+
+pub use pact_matching::Mismatch;
+
+ffi_fn! {
+    /// Match a pair of messages, producing a collection of mismatches,
+    /// which is empty if the two messages matched.
+    fn match_message(msg_1: *const Message, msg_2: *const Message) -> *const Mismatches {
+        let msg_1 = as_ref!(msg_1);
+        let msg_2 = as_ref!(msg_2);
+        let mismatches = Mismatches(pm::match_message(msg_1, msg_2));
+
+        ptr::raw_to(mismatches) as *const Mismatches
+    } {
+        ptr::null_to::<Mismatches>() as *const Mismatches
+    }
+}
+
+ffi_fn! {
+    /// Get an iterator over mismatches.
+    fn mismatches_get_iter(mismatches: *const Mismatches) -> *mut MismatchesIterator {
+        let mismatches = as_ref!(mismatches);
+        let iter = MismatchesIterator { current: 0, mismatches };
+        ptr::raw_to(iter)
+    } {
+        ptr::null_mut_to::<MismatchesIterator>()
+    }
+}
+
+ffi_fn! {
+    /// Delete mismatches
+    fn mismatches_delete(mismatches: *const Mismatches) {
+        ptr::drop_raw(mismatches as *mut Mismatches);
+    }
+}
+
+ffi_fn! {
+    /// Get the next mismatch from a mismatches iterator.
+    ///
+    /// Returns a null pointer if no mismatches remain.
+    fn mismatches_iter_next(iter: *mut MismatchesIterator) -> *const Mismatch {
+        let iter = as_mut!(iter);
+        let mismatches = as_ref!(iter.mismatches);
+        let index = iter.next();
+        let mismatch = mismatches
+            .0
+            .get(index)
+            .ok_or(anyhow::anyhow!("iter past the end of mismatches"))?;
+       mismatch as *const Mismatch
+    } {
+        ptr::null_to::<Mismatch>()
+    }
+}
+
+ffi_fn! {
+    /// Delete a mismatches iterator when you're done with it.
+    fn mismatches_iter_delete(iter: *mut MismatchesIterator) {
+        ptr::drop_raw(iter);
+    }
+}
+
+ffi_fn! {
+    /// Get a JSON representation of the mismatch.
+    fn mismatch_to_json(mismatch: *const Mismatch) -> *const c_char {
+        let mismatch = as_ref!(mismatch);
+        let json = mismatch.to_json().to_string();
+        string::to_c(&json)? as *const c_char
+    } {
+        ptr::null_to::<c_char>()
+    }
+}
+
+ffi_fn! {
+    /// Get the type of a mismtatch.
+    fn mismatch_type(mismatch: *const Mismatch) -> *const c_char {
```

Comment:
So the scope of mismatch types is wide enough that we need to fall back on a raw string instead of something like an enum?  Just clarifying.

Reply:
This is for consistency with the underlying implementation, and simplicity in the FFI API. Right now, `pact_matching` returns a string from `Mismatch::mismatch_type`, so what we're doing here is simply getting that string and concerting it into a C-compatible representation. We _could_, in the FFI, define an enum which matches against the string from the Rust side, but it'd be brittle if the underlying `pact_matching` implementation decided to add new variants. I'm not opposed to it, but that's my perspective right now.

Reply:
OK - I'm good with that then!

Reply:
Great! I'll wait for @cstepanian to give a review too before merging. Thanks!

Reply:
This looks fine to me --- this code is just exposing the available methods on the underlying `Mismatch` enum. I imagine this particular method should only be used for human-facing text output, and a client wanting the structure of the Mismatch could instead call `mismatch_to_json` and parse the JSON.
In the future, perhaps another interface could be added that gives the calling code a direct view of the enum, without having to go through JSON.

And while we're talking about this bit of code:
On line 89, it appears that "mismatch" is misspelled.

Reply:
Alright, I've fixed the typo. Thanks @cstepanian!